llvm/tools/sycl-post-link/ModuleSplitter.h

```diff
@@ -34,66 +35,122 @@ enum IRSplitMode {
-  EntryPointGroup(StringRef GroupId, EntryPointVec &&Functions)
+  EntryPointGroup(StringRef GroupId, EntryPointSet &&Functions)
       : GroupId(GroupId), Functions(std::move(Functions)) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet &&Functions,
+                  const Properties &Props)
+      : GroupId(GroupId), Functions(std::move(Functions)), Props(Props) {}
+
+  // Tells if this group has only ESIMD entry points.
+  bool isEsimd() const {
+    return Props.HasESIMD == SyclEsimdSplitStatus::ESIMD_ONLY;
+  }
+  // Tells if this group has only SYCL entry points.
+  bool isSycl() const {
+    return Props.HasESIMD == SyclEsimdSplitStatus::SYCL_ONLY;
+  }
+  // Tells if some entry points use double GRF mode.
+  bool isDoubleGRF() const { return Props.UsesDoubleGRF; }
 
-  const StringRef getId() const { return GroupId; }
-  // Tells if this group has only ESIMD entry points (based on GroupId).
-  bool isEsimd() const;
-  // Tells if this group has only SYCL entry points (based on GroupId).
-  bool isSycl() const;
   void saveNames(std::vector<std::string> &Dest) const;
   void rebuildFromNames(const std::vector<std::string> &Names, const Module &M);
   void rebuild(const Module &M);
 };
 
 using EntryPointGroupVec = std::vector<EntryPointGroup>;
 
-enum class SyclEsimdSplitStatus { SYCL_ONLY, ESIMD_ONLY, SYCL_AND_ESIMD };
-
+// Annotates an llvm::Module with information necessary to perform and track
+// result of device code (llvm::Module instances) splitting:
+// - entry points of the module determined e.g. by a module splitter, as well
+//   as information about entry point origin (e.g. result of a scoped split)
+// - its properties, such as whether it has specialization constants uses
+// It also provides convenience functions for entry point set transformation
+// between llvm::Function object and string representations.
 class ModuleDesc {
   std::unique_ptr<Module> M;
   EntryPointGroup EntryPoints;
+  bool IsTopLevel = false;
 
 public:
   struct Properties {
-    SyclEsimdSplitStatus HasEsimd = SyclEsimdSplitStatus::SYCL_AND_ESIMD;
-    bool SpecConstsMet = true;
+    bool SpecConstsMet = false;
   };
   std::string Name = "";
   Properties Props;
 
-  ModuleDesc(std::unique_ptr<Module> &&M, EntryPointGroup &&EntryPoints)
-      : M(std::move(M)), EntryPoints(std::move(EntryPoints)) {
-    Name = this->EntryPoints.getId().str();
+  ModuleDesc(std::unique_ptr<Module> &&M, StringRef Name = "TOP-LEVEL")
+      : M(std::move(M)), IsTopLevel(true), Name(Name) {}
+
+  ModuleDesc(std::unique_ptr<Module> &&M, EntryPointGroup &&EntryPoints,
+             const Properties &Props)
+      : M(std::move(M)), EntryPoints(std::move(EntryPoints)), Props(Props) {
+    Name = this->EntryPoints.GroupId.str();
   }
 
-  ModuleDesc(std::unique_ptr<Module> &&M, const std::vector<std::string> &Names)
-      : M(std::move(M)) {
+  ModuleDesc(std::unique_ptr<Module> &&M, const std::vector<std::string> &Names,
+             StringRef Name = "NoName")
```

Comment:
Why is this initializer value different from that used in Line 122?

Reply:
I see it now. In Line 122, it is a top-level module and hence the name is initialize to TOP-LEVEL. In Line 132, the module 'level' is not known. Hence, a dummy "NoName" is used. I think we will eventually get rid of these names, if I remember @kbobrovs comments.